proxy/limits/limits.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !windows
+
+// Package limits provides routines to check and enforce certain resource
+// limits on the Cloud SQL client proxy process.
+package limits
+
+import (
+	"fmt"
+	"syscall"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+)
+
+var (
+	// For overriding in unittests.
+	syscallGetrlimit = syscall.Getrlimit
+	syscallSetrlimit = syscall.Setrlimit
+)
+
+// SetupFDLimits ensures that the process running the Cloud SQL proxy can have
+// at least wantFDs number of open file descriptors. It returns an error if it
+// cannot ensure the same.
+func SetupFDLimits(wantFDs uint64) error {
+	curr := &syscall.Rlimit{}
+	if err := syscallGetrlimit(syscall.RLIMIT_NOFILE, curr); err != nil {
+		return fmt.Errorf("failed to read rlimit for max file descriptors: %v", err)
+	}
+
+	if curr.Max < wantFDs {
```

Comment:
From my read of the manpage Max may be RLIMIT_INFINITY, which can be a negative constant. I think to be safe we should explicitly check for that special case.

Reply:
type Rlimit struct {
        Cur uint64
        Max uint64
}
RLIM_INFINITY I believe would be -1, which will show up as math.MaxUint64. So, our checks here should be unaffected. Am I missing something.

Reply:
Also, I changed the logic to try and set the rlimit even if the current hard limit is less than the value passed in the flag, since the manpage claims that processes with sufficient privileges can override the hard limit as well. PTAL.